perf_benchmark/benchmark_common.cc

```diff
@@ -0,0 +1,294 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+//
+#include <fstream>
+#include <sstream>
+#include <utility>
+#include "perf_benchmark/benchmark_common.h"
+#include "google/protobuf/text_format.h"
+#include "absl/strings/escaping.h"
+#include "nlohmann/json.hpp"
+
+namespace google {
+namespace grpc {
+namespace transcoding {
+
+namespace perf_benchmark {
+
+namespace pb = ::google::protobuf;
+
+std::string LoadFile(const std::string& file_name) {
+  std::ifstream ifs(file_name);
+  if (!ifs) {
+    std::cerr << "Could not open " << file_name.c_str() << std::endl;
+    return {};
+  }
+  std::ostringstream ss;
+  ss << ifs.rdbuf();
+  return ss.str();
+}
+
+bool LoadService(const std::string& config_pb_txt_file,
+                 ::google::api::Service* service) {
+  static const char kBenchmarkData[] = "perf_benchmark/";
+  return LoadService(config_pb_txt_file, kBenchmarkData, service);
+}
+
+bool LoadService(const std::string& config_pb_txt_file,
+                 const std::string& benchmark_path,
+                 ::google::api::Service* service) {
+  auto config = LoadFile(benchmark_path + config_pb_txt_file);
+  if (config.empty()) {
+    return false;
+  }
+
+  if (!pb::TextFormat::ParseFromString(config, service)) {
+    std::cerr << "Could not parse service config from "
+              << config_pb_txt_file.c_str() << std::endl;
+    return false;
+  } else {
+    return true;
+  }
+}
+double GetPercentile(const std::vector<double>& v, double perc) {
+  if (perc < 0) { perc = 0; }
+  if (perc > 100) { perc = 100; }
+  // Making a copy since std::nth_element mutates the vector
+  auto copy = std::vector<double>(v);
+  size_t rough_position = copy.size() * perc / 100;
+  std::nth_element(copy.begin(), copy.begin() + rough_position, copy.end());
+  return copy[rough_position];
+}
+
+std::string GetRandomBytesString(int64_t length, bool base64) {
+  static const char charset[] =
```

Comment:
Reader has no idea which ascii characters these are :)

Is there any utility you can use in absl to make these more readable? https://github.com/abseil/abseil-cpp/blob/master/absl/strings/ascii.h

Reply:
Yeah, it is hard to read. But the charset covers every ASCII characters. The charset elements range from 0 to 127.
Do you think renaming the `charset` variable to `full_ascii_charset` would be better?

Reply:
See my other comment below on how to simplify this.